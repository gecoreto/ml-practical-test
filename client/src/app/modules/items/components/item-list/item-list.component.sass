@import 'utils'

.item-list-container
  background: var(--ml-white-color)
  .product-item:not(:last-child)
    border-bottom: thin solid #eee

  .product-item
    cursor: pointer
    padding: .5rem 0
    &-img
      .item-image
        width: 100%

    &-info
      &-options
        color: #333
        &.amount
          .price
            display: inline-block
            margin-right: .5rem
            margin-bottom: 0
          .delivery-icon
            background-color: green
            border-radius: 50%
            padding: 2px
            > img
              width: 15px
              height: 15px
              position: relative
              top: 2px

  @include media-width-min(md)
    .product-item
      display: flex
      &-img
        .item-image
          width: 160px
      &-info
        width: 100%
        &-options
          &.location
            color: #666666
            float: right
            margin: 32px 32px 0 0
