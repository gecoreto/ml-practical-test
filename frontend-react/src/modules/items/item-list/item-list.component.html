<div class="item-list-container p-1">
    <ng-container *ngIf="(working$ | async); else ((items$ | async).length > 0) ? content : empty">
        <ng-container *ngFor="let item of [1,1,1,1]">
            <div class="card-loader card-loader-item-1"></div>
        </ng-container>
    </ng-container>
    <ng-template #content>
        <div
            *ngFor="let item of (items$ | async)"
            [routerLink]="['/items', item.id]"
            class="product-item">
            <div class="product-item-img">
                <figure>
                    <img
                        [src]="item.picture"
                        class="item-image"
                        [alt]="item.title">
                </figure>
            </div>
            <div class="product-item-info">
                <div class="product-item-info-options location">{{ item.city_name | titlecase }}</div>
                <div class="product-item-info-options amount">
                    <h2 class="price">
                        {{ item.price.amount | currency:item.price.amount.currency }}
                    </h2>
                    <span
                        *ngIf="item.free_shipping"
                        class="delivery-icon">
                        <img
                            src="./assets/img/track.svg"
                            alt="Envio">
                    </span>
                </div>
                <div class="product-item-info-options">
                    <h3 class="m-0">{{ item.title }}</h3>
                </div>

            </div>
        </div>
    </ng-template>
    <ng-template #empty>Bienvenido</ng-template>
</div>
