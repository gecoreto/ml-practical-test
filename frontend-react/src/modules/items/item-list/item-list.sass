@import '../../../sass/utils'

.item-list-container
  background: var(--ml-white-color)
  .product-item:not(:last-child)
    border-bottom: thin solid #eee

  .product-item
    cursor: pointer
    padding: 1rem 0
    &-img
      .item-image
        width: 100%

    &-info
      &-options
        color: #333
        position: relative
        > h3
          font-size: 1.125rem
        &.amount
          margin-bottom: 1rem
          .price
            display: inline-block
            font-size: 1.5rem
            margin: 0 8px 0 0
          .delivery-icon
            background-color: green
            border-radius: 50%
            padding: 2px
            position: relative
            top: -3px
            > img
              width: 15px
              height: 15px
              position: relative
              top: 2px
        .location
          font-size: 12px
          color: #666666
          position: absolute
          bottom: 8px
          right: 0

  @include media-width-min(md)
    .product-item
      display: flex
      &-img
        .item-image
          width: 180px
          height: 180px
          margin-right: 16px
      &-info
        width: 100%
        &-options
          position: relative
          .location
            bottom: 0
            margin-right: 32px
