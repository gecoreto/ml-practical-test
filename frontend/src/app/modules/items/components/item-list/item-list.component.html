<div class="item-list-container p-1">
    <ng-container *ngIf="(working$ | async); else ((items$ | async).length > 0) ? content : empty">
        <ng-container *ngFor="let item of [1,1,1,1]">
            <div class="card-loader card-loader-item-1"></div>
        </ng-container>
    </ng-container>
    <ng-template #content>
        <pre>
            {{  (items$ | async) | json }}
        </pre>
        <div
            *ngFor="let item of [1,1,1]"
            [routerLink]="['/items', 2]"
            class="product-item">
            <div class="product-item-img">
                <figure>
                    <img
                        src="https://http2.mlstatic.com/D_NQ_NP_805950-MCO40352919808_012020-O.webp"
                        class="item-image"
                        alt="Celular iPhone 11 128gb Nuevo 100% Original Y Sellada ">
                </figure>
            </div>
            <div class="product-item-info">
                <div class="product-item-info-options location">Bogot√°</div>
                <div class="product-item-info-options amount">
                    <h2 class="price">$ 300.000</h2>
                    <span class="delivery-icon">
                        <img
                            src="./assets/img/track.svg"
                            alt="Envio">
                    </span>
                </div>
                <div class="product-item-info-options">
                    <h3 class="m-0">
                        Celular iPhone 11 128gb Nuevo 100% Original Y Sellada
                    </h3>
                </div>
                
            </div>
        </div>
    </ng-template>
    <ng-template #empty>
        Bienvenido
    </ng-template>
</div>
