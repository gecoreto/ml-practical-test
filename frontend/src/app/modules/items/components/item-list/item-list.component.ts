import { filter, tap } from 'rxjs/operators';
import { ItemsI } from './../../entities/items.entity';
import { Observable } from 'rxjs';
import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ItemsFacade } from './../../items.facade';

@Component({
  selector: 'app-item-list',
  templateUrl: './item-list.component.html',
  styleUrls: ['./item-list.component.sass'],
})
export class ItemListComponent implements OnInit {
  constructor(private route: ActivatedRoute, private facade: ItemsFacade) {}

  ngOnInit(): void {
    this.route.queryParams
      .pipe(
        tap(() => this.facade.itemsClearStore()),
        filter(({ search }) => !!search)
      )
      .subscribe(({ search }) => this.facade.fetchSearchProducts(search));
  }

  get items$(): Observable<ItemsI[]> {
    return this.facade.items$;
  }

  get working$(): Observable<boolean> {
    return this.facade.working$;
  }
}
