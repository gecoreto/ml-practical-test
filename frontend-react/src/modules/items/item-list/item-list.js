import { bindActionCreators } from '@reduxjs/toolkit';
import React from 'react';
import { connect } from 'react-redux';
import trackImg from '../../../assets/img/track.svg';
import currencyFormat from '../../../utils/format';
import { fetchSearchProducts, selectItems, selectWorking } from '../store/itemsSlice';
import './item-list.sass';

const Loading = () => [1, 2, 3, 4].map((number) =>
  <div key={number.toString()} className="card-loader card-loader-item-1"></div>
)

const DeliveryIcon = () => (
  <span className="delivery-icon">
    <img
      src={trackImg}
      alt="Envio" />
  </span>
)

class Itemlist extends React.Component {

  componentDidMount() {
    const query = new URLSearchParams(this.props.location.search)
    this.fetchProducts(query.get('search'));
  }

  fetchProducts(search) {
    if (!!search) {
      this.props.fetchSearchProducts(search);
    }
  }

  goToDetail(url) {
    let history = this.props.history;
    history.push(url);
  }

  render() {
    const { items, working } = this.props;
    let { url } = this.props.match;

    return working
      ? <Loading />
      : (
        <div className="item-list-container p-1">
          {(items.length === 0)
            ? 'Bienvenido'
            : items.map((item) =>
              <div id={item.id.toString()} key={item.id.toString()} className="product-item" onClick={() => this.goToDetail(`${url}/${item.id}`)}>
                <div className="product-item-img">
                  <figure>
                    <img
                      src={item.picture}
                      className="item-image"
                      alt={item.title} />
                  </figure>
                </div>
                <div className="product-item-info">
                  <div className="product-item-info-options location">{item.city_name}</div>
                  <div className="product-item-info-options amount">
                    <h2 className="price">
                      {/* {{ item.price.amount | currency:item.price.amount.currency }} */}
                      {currencyFormat(item.price.amount)}
                    </h2>
                    {item.free_shipping ? <DeliveryIcon /> : ''}
                  </div>
                  <div className="product-item-info-options">
                    <h3 className="m-0">{item.title}</h3>
                  </div>
                </div>
              </div>
            )}
        </div>
      );
  }
}

function mapStateToProps(state) {
  return {
    items: selectItems(state),
    working: selectWorking(state)
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ fetchSearchProducts }, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Itemlist)

export { Itemlist };