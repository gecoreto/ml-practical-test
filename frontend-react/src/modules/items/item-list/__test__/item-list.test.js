import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import React from 'react';
import { DeliveryIcon, Itemlist, Loading } from '../item-list';

const mockHistoryPush = jest.fn();
const mockFetchSearchProducts = jest.fn();

describe("Itemlist component", () => {

    const props = {
        match: { url: '' },
        items: [
            {
                id: "MLA916044386",
                title: "Ps5",
                installments: { currency_id: "ARS" },
                price: { currency: "ARS", amount: 5000 },
                address: { city_name: "Mar de la plata" },
                free_shipping: true
            },
        ],
        working: false,
        location: { search: '?search=ps5' },
        history: {
            push: mockHistoryPush
        },
        fetchSearchProducts: mockFetchSearchProducts
    };

    const wrapper = shallow(<Itemlist {...props} />);

    test('should test Itemlist component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe("WHEN is clicked", () => {
        const id = props.items[0].id;
        it("THEN go To Detail", () => {
            wrapper.find(`[id="${id}"]`).simulate("click");
            expect(mockHistoryPush).toHaveBeenCalledWith(`/${id}`);
        });
    });

    describe("WHEN It is loading", () => {
        test('renders Loading', () => {
            const { container } = render(<Loading />);
            expect(container.getElementsByClassName('card-loader').length).toBe(4);
        });
    });

    describe("WHEN It is DeliveryIcon", () => {
        test('renders DeliveryIcon', () => {
            const { container } = render(<DeliveryIcon />);
            expect(container.getElementsByClassName('delivery-icon').length).toBe(1);
        });
    });

});
