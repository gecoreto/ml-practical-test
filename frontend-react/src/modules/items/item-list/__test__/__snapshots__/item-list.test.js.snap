// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Itemlist component should test Itemlist component 1`] = `ShallowWrapper {}`;
